libsolidity/formal/SymbolicState.cpp

```diff
@@ -149,6 +153,120 @@ smtutil::Expression SymbolicState::txConstraints(FunctionDefinition const& _func
 	return conj;
 }
 
+void SymbolicState::buildABIFunctions(set<FunctionCall const*> const& _abiFunctions)
+{
+	map<string, SortPointer> functions;
+
+	for (auto const* funCall: _abiFunctions)
+	{
+		auto t = dynamic_cast<FunctionType const*>(funCall->expression().annotation().type);
+
+		auto const& args = funCall->sortedArguments();
+		auto const& paramTypes = t->parameterTypes();
+		auto const& returnTypes = t->returnParameterTypes();
+
+
+		auto argTypes = [](auto const& _args) {
+			return applyMap(_args, [](auto arg) { return arg->annotation().type; });
+		};
+
+		vector<TypePointer> inTypes;
+		vector<TypePointer> outTypes;
+		if (t->kind() == FunctionType::Kind::ABIDecode)
+		{
+			inTypes.emplace_back(TypeProvider::bytesMemory());
+			auto const* tupleType = dynamic_cast<TupleType const*>(args.at(1)->annotation().type);
+			solAssert(tupleType, "");
+			for (auto t: tupleType->components())
+			{
+				auto typeType = dynamic_cast<TypeType const*>(t);
+				solAssert(typeType, "");
+				outTypes.emplace_back(typeType->actualType());
+			}
+		}
+		else
+		{
+			outTypes = returnTypes;
+			if (
+				t->kind() == FunctionType::Kind::ABIEncodeWithSelector ||
+				t->kind() == FunctionType::Kind::ABIEncodeWithSignature
+			)
+			{
+				inTypes.emplace_back(paramTypes.front());
+				inTypes += argTypes(vector<ASTPointer<Expression const>>(args.begin() + 1, args.end()));
+			}
+			else
+			{
+				solAssert(
+					t->kind() == FunctionType::Kind::ABIEncode ||
+					t->kind() == FunctionType::Kind::ABIEncodePacked,
+					""
+				);
+				inTypes = argTypes(args);
+			}
+		}
+
+		auto replaceTypes = [](auto& _types) {
+			for (auto& t: _types)
+				if (t->category() == frontend::Type::Category::RationalNumber)
+					t = TypeProvider::uint256();
+				else if (t->category() == frontend::Type::Category::StringLiteral)
+					t = TypeProvider::bytesMemory();
+		};
+		replaceTypes(inTypes);
+		replaceTypes(outTypes);
+
+		auto name = t->richIdentifier();
+		for (auto in: inTypes)
+			name += "_" + in->toString(true);
+		for (auto out: outTypes)
+			name += "_" + out->toString(true);
+
+		m_abiMembers[funCall] = {name, inTypes, outTypes};
+
+		if (functions.count(name))
+			continue;
+
+		auto typesToSort = [](auto const& _types, string const& _name) -> shared_ptr<Sort> {
+			if (_types.size() == 1)
+				return smtSort(*_types.front());
+
+			vector<string> inNames;
+			vector<SortPointer> sorts;
+			for (unsigned i = 0; i < _types.size(); ++i)
+			{
+				inNames.emplace_back(_name + "_input_" + to_string(i));
+				sorts.emplace_back(smtSort(*_types.at(i)));
+			}
+			return make_shared<smtutil::TupleSort>(
+				_name + "_input",
+				inNames,
```

Comment:
Needing to define names for all of these is a bit annoying...

Reply:
It is... but those names are used as accessor functions for the tuples in SMT

Reply:
I'm aware, just voicing my discontent about the fact that tuples work like that in SMT - but yeah - basically s spam comment, sorry ;-).